src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1414 @@
+    }
+
+    partial class BoundAssignmentOperator : IAssignment
+    {
+        IReference IAssignment.Target => this.Left as IReference;
+
+        IExpression IAssignment.Value => this.Right;
+
+        protected override OperationKind ExpressionKind => OperationKind.Assignment;
+    }
+
+    partial class BoundCompoundAssignmentOperator : ICompoundAssignment
+    {
+        BinaryOperationKind ICompoundAssignment.BinaryKind => Expression.DeriveBinaryOperationKind(this.Operator.Kind);
+
+        IReference IAssignment.Target => this.Left as IReference;
+
+        IExpression IAssignment.Value => this.Right;
+
+        bool IWithOperator.UsesOperatorMethod => (this.Operator.Kind & BinaryOperatorKind.TypeMask) == BinaryOperatorKind.UserDefined;
+
+        IMethodSymbol IWithOperator.Operator => this.Operator.Method;
+
+        protected override OperationKind ExpressionKind => OperationKind.CompoundAssignment;
+    }
+
+    partial class BoundIncrementOperator : IIncrement
+    {
+        UnaryOperationKind IIncrement.IncrementKind => Expression.DeriveUnaryOperationKind(this.OperatorKind);
+
+        BinaryOperationKind ICompoundAssignment.BinaryKind => Expression.DeriveBinaryOperationKind(((IIncrement)this).IncrementKind);
+
+        IReference IAssignment.Target => this.Operand as IReference;
+
+        private static readonly System.Runtime.CompilerServices.ConditionalWeakTable<BoundIncrementOperator, IExpression> IncrementValueMappings = new System.Runtime.CompilerServices.ConditionalWeakTable<BoundIncrementOperator, IExpression>();
+
+        IExpression IAssignment.Value => IncrementValueMappings.GetValue(this, (increment) => new BoundLiteral(this.Syntax, Semantics.Expression.SynthesizeNumeric(increment.Type, 1), increment.Type));
+
+        bool IWithOperator.UsesOperatorMethod => (this.OperatorKind & UnaryOperatorKind.TypeMask) == UnaryOperatorKind.UserDefined;
+
+        IMethodSymbol IWithOperator.Operator => this.MethodOpt;
+
+        protected override OperationKind ExpressionKind => OperationKind.Increment;
+    }
+
+    partial class BoundBadExpression
+    {
+        protected override OperationKind ExpressionKind => OperationKind.None;
+    }
+
+    partial class BoundNewT
+    {
+        protected override OperationKind ExpressionKind => OperationKind.TypeParameterObjectCreation;
+    }
+
+    partial class BoundUnaryOperator : IUnary
+    {
+        UnaryOperationKind IUnary.UnaryKind => Expression.DeriveUnaryOperationKind(this.OperatorKind);
+
+        IExpression IUnary.Operand => this.Operand;
+
+        bool IWithOperator.UsesOperatorMethod => (this.OperatorKind & UnaryOperatorKind.TypeMask) == UnaryOperatorKind.UserDefined;
+
+        IMethodSymbol IWithOperator.Operator => this.MethodOpt;
+
+        protected override OperationKind ExpressionKind => OperationKind.UnaryOperator;
+    }
+
+    partial class BoundBinaryOperator : IBinary, IRelational
```

Comment:
Conceptually, yes. I expect to yield to pressure and eliminate IRelational, though.